Client/Frontend/Browser/MetadataParserHelper.swift

```diff
@@ -3,50 +3,102 @@
-    private weak var tab: Tab?
-    private let profile: Profile
+class MetadataParserHelper: TabEventHandler {
+    private var tabObservers: TabObservers!
 
-    class func name() -> String {
-        return "MetadataParserHelper"
+    init() {
+        self.tabObservers = registerFor(
+            .didChangeURL,
+            queue: .main)
     }
 
-    required init(tab: Tab, profile: Profile) {
-        self.tab = tab
-        self.profile = profile
+    deinit {
+        unregister(tabObservers)
     }
 
-    func scriptMessageHandlerName() -> String? {
-        return "metadataMessageHandler"
-    }
-
-    func userContentController(_ userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
+    func tab(_ tab: Tab, didChangeURL url: URL) {
         // Get the metadata out of the page-metadata-parser, and into a type safe struct as soon
         // as possible.
-        guard let dict = message.body as? [String: Any],
-            let tab = self.tab,
-            let pageURL = tab.url?.displayURL,
-            let pageMetadata = PageMetadata.fromDictionary(dict) else {
-                log.debug("Page contains no metadata!")
+        guard let webView = tab.webView,
+            let url = webView.url, url.isWebPage(includeDataURIs: false), !url.isLocal else {
+            return
+        }
+
+        webView.evaluateJavaScript("__firefox__.metadata && __firefox__.metadata.getMetadata()") { (result, error) in
+            guard error == nil else {
                 return
+            }
+
+            guard let dict = result as? [String: Any],
+                let pageURL = tab.url?.displayURL,
+                let pageMetadata = PageMetadata.fromDictionary(dict) else {
+                    log.debug("Page contains no metadata!")
+                    return
+            }
+
+            tab.pageMetadata = pageMetadata
+            TabEvent.post(.didLoadPageMetadata(pageMetadata), for: tab)
+
+            let userInfo: [String: Any] = [
+                "isPrivate": tab.isPrivate,
+                "pageMetadata": pageMetadata,
+                "tabURL": pageURL
+            ]
+            NotificationCenter.default.post(name: .OnPageMetadataFetched, object: nil, userInfo: userInfo)
         }
+    }
+}
 
-        let userInfo: [String: Any] = [
-            "isPrivate": self.tab?.isPrivate ?? true,
-            "pageMetadata": pageMetadata,
-            "tabURL": pageURL
-        ]
+class MediaImageLoader: TabEventHandler {
+    private var tabObservers: TabObservers!
+    private let prefs: Prefs
+
+    init(_ prefs: Prefs) {
+        self.prefs = prefs
+        self.tabObservers = registerFor(
+            .didLoadPageMetadata,
+            queue: .main)
+    }
 
-        tab.pageMetadata = pageMetadata
+    deinit {
+        unregister(tabObservers)
+    }
+
+    func tab(_ tab: Tab, didLoadPageMetadata metadata: PageMetadata) {
+        let cacheImages = !NoImageModeHelper.isActivated(prefs)
```

Comment:
I actually don't think we should check this. These images are used to improve AS and I don't think should have any correlation to whether or not "no image mode" is enabled for web content. We had no choice but to tie the two together before.

Reply:
After discussion in slack, we don't know if this No Image Mode is the same as Reduced Data Mode.

This is the same as previous behaviour, so leaving it here.